### PR TITLE
Do not create new addrConn when connection error happens

### PR DESCRIPTION
**Pull request by menghanl**
Just reset the underlying transport instead.

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -936,28 +950,46 @@ func (ac *addrConn) transportMonitor() {
...
-			// In both cases, a new ac is created.
+			// If GoAway happens without any network I/O error, the underlying transport
+			// will be gracefully closed, and a new transport will be created.
+			// (the transport will be closed when all the pending RPCs finished or failed.).
+			// If GoAway and some network I/O error happen concurrently, the underlying transport
+			// will be closed, and a new transport will be created.
 			select {
 			case <-t.Error():
-				ac.cc.resetAddrConn(ac.addr, false, errNetworkIO)
+				if err := ac.resetTransport(false); err != nil {
```
Comment (dfawley) [inline clientconn.go]: Refactor?

```go
var drain bool
select {
  case <-t.Error():
  default:
    drain = true
}
if err := ac.resetTransport(drain); err != nil {
...
```

Comment (menghanl) [inline clientconn.go]: Done. PTAL.

File: clientconn.go [hunk 2]
```diff
@@ -599,10 +599,7 @@ func (cc *ClientConn) resetAddrConn(addr Address, block bool, tearDownErr error)
 	cc.mu.Unlock()
 	if stale != nil {
 		// There is an addrConn alive on ac.addr already. This could be due to
-		// 1) a buggy Balancer notifies duplicated Addresses;
-		// 2) goaway was received, a new ac will replace the old ac.
-		//    The old ac should be deleted from cc.conns, but the
-		//    underlying transport should drain rather than close.
+		// a buggy Balancer notifies duplicated Addresses;
```
Comment (dfawley) [inline clientconn.go]: "a buggy Balancer that reports duplicate Addresses."

File: clientconn.go [hunk 3]
```diff
@@ -949,28 +949,22 @@ func (ac *addrConn) transportMonitor() {
 			ac.adjustParams(t.GetGoAwayReason())
 			// If GoAway happens without any network I/O error, the underlying transport
 			// will be gracefully closed, and a new transport will be created.
-			// (the transport will be closed when all the pending RPCs finished or failed.).
+			// (the transport will be closed when all the pending RPCs finished or failed.)
```
Comment (dfawley) [inline clientconn.go]: Capital "T" at the start of the sentence.

**Pull request merged**